code/modules/antagonists/traitor/objectives/sabotage_machinery.dm

```diff
@@ -0,0 +1,235 @@
+
+/// Traitor objective to place a trap on a machine which causes it to detonate
+/datum/traitor_objective_category/sabotage_machinery
+	name = "Sabotage Worksite"
+	objectives = list(
+		/datum/traitor_objective/sabotage_machinery/trap = 1,
+		/datum/traitor_objective/sabotage_machinery/destroy = 1,
+	)
+
+/datum/traitor_objective/sabotage_machinery
+	name = "Sabotage the %MACHINE%"
+	description = "Abstract objective holder which shouldn't appear in your uplink."
+
+	progression_reward = list(2 MINUTES, 8 MINUTES)
+	telecrystal_reward = list(0, 1)
+
+	progression_minimum = 0 MINUTES
+	progression_maximum = 10 MINUTES
+
+	abstract_type = /datum/traitor_objective/sabotage_machinery
+
+	/// The maximum amount of this type of objective a traitor can have, set to 0 for no limit.
+	var/maximum_allowed = 0
+	/// The possible target machinery and the jobs tied to each one.
+	var/list/applicable_jobs = list()
+	/// The chosen job. Used to check for duplicates
+	var/chosen_job
+
+/datum/traitor_objective/sabotage_machinery/can_generate_objective(datum/mind/generating_for, list/possible_duplicates)
+	if(!maximum_allowed)
+		return TRUE
+	if(length(possible_duplicates) >= maximum_allowed)
+		return FALSE
+	return TRUE
+
+/datum/traitor_objective/sabotage_machinery/generate_objective(datum/mind/generating_for, list/possible_duplicates)
+	var/list/possible_jobs = applicable_jobs.Copy()
+	for(var/datum/traitor_objective/sabotage_machinery/objective as anything in possible_duplicates)
+		possible_jobs -= objective.chosen_job
+	for(var/available_job in possible_jobs)
+		var/job_machine_path = possible_jobs[available_job]
+		if (!length(GLOB.objective_machine_handler.machine_instances_by_path[job_machine_path]))
+			possible_jobs -= available_job
+	if(!length(possible_jobs))
+		return FALSE
+
+	chosen_job = pick(possible_jobs)
+	var/list/obj/machinery/possible_machines = GLOB.objective_machine_handler.machine_instances_by_path[possible_jobs[chosen_job]]
+	for(var/obj/machinery/machine as anything in possible_machines)
+		prepare_machine(machine)
+
+	replace_in_name("%JOB%", lowertext(chosen_job))
+	replace_in_name("%MACHINE%", possible_machines[1].name)
+	return TRUE
+
+/// Marks a given machine as our target
+/datum/traitor_objective/sabotage_machinery/proc/prepare_machine(obj/machinery/machine)
+	AddComponent(/datum/component/traitor_objective_register, machine, succeed_signals = list(COMSIG_PARENT_QDELETING))
+
+/datum/traitor_objective/sabotage_machinery/destroy
+	name = "Destroy the %MACHINE%"
+	description = "Destroy the %MACHINE% to cause disarray and disrupt the operations of the %JOB%'s department."
+
+	progression_reward = list(5 MINUTES, 10 MINUTES)
+	telecrystal_reward = list(3, 4)
+
+	progression_minimum = 15 MINUTES
+	progression_maximum = 30 MINUTES
+
+	applicable_jobs = list(
+		JOB_STATION_ENGINEER = /obj/machinery/telecomms/hub,
+		JOB_SCIENTIST = /obj/machinery/rnd/server,
+	)
+
+/datum/traitor_objective/sabotage_machinery/trap
```

Comment:
You should explicitly set the telecrystal reward and progression for each non-abstract objective so that it's clear how much they award. Abstract types should not have a telecrystal reward and progression set so that the unit test fails when an objective doesn't explicitly set their own telecrystal reward and progression.

Reply:
I'll be honest I was kind of thinking that when I did it, so this makes sense to me.